src/components/forms/FloatingWidget.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * Implements a form field widget that floats in a 'popout'.
+ *
+ * Focusing the form field opens the widget, while keeping focus on original form
+ * field. Keyboard tab navigation focuses the elements inside the popout in order,
+ * while any other keyboard input closes the widget (shifting to 'manual input' mode).
+ *
+ * Clicking the form field/reference (re-)opens the widget. The widget/popout can be
+ * dismissed by hitting the [ESC} key, clicking outside of it or tabbing through/out
+ * of it.
+ *
+ * You need both the hook and component for the behaviour.
+ *
+ * Note that `getReferenceProps` returns event handlers like onKeyDown - if you are
+ * overriding these events yourself, make sure you don't accidentally overwrite the
+ * floating widget handlers.
+ */
+import {
+  FloatingArrow,
+  FloatingFocusManager,
+  arrow,
+  autoUpdate,
+  flip,
+  offset,
+  shift,
+  useDismiss,
+  useFloating,
+  useInteractions,
+  useRole,
+} from '@floating-ui/react';
+import PropTypes from 'prop-types';
+import React, {useRef, useState} from 'react';
+
+import {getBEMClassName} from 'utils';
+
+const useFloatingWidget = () => {
+  const [isOpen, setIsOpen] = useState(false);
+  const keepClosed = useRef(false);
+  const arrowRef = useRef(null);
+
+  const manageOpen = (open, options = {}) => {
+    const {keepDismissed = false} = options;
+    setIsOpen(open);
+    if (open && keepClosed.current) {
+      keepClosed.current = false;
+    }
+    if (!open && keepDismissed) {
+      keepClosed.current = true;
+    }
+  };
+
+  const {refs, floatingStyles, context} = useFloating({
+    open: isOpen,
+    onOpenChange: manageOpen,
+    middleware: [offset(10), flip(), shift(), arrow({element: arrowRef})],
+    whileElementsMounted: autoUpdate,
+  });
+
+  const dismiss = useDismiss(context);
+  const role = useRole(context);
+
+  const {getReferenceProps, getFloatingProps} = useInteractions([dismiss, role]);
+  const referenceProps = getReferenceProps();
+
+  const onClick = () => {
+    const isFocused = context.refs.reference.current === document.activeElement;
+    isFocused && !isOpen && manageOpen(true);
+  };
+
+  const onKeyDown = event => {
+    referenceProps?.onKeyDown?.(event);
+    event.key !== 'Tab' && manageOpen(false);
+  };
+
+  const onFocus = () => {
+    !isOpen && !keepClosed.current && manageOpen(true);
+    if (keepClosed.current) {
+      keepClosed.current = false;
+    }
+  };
+
+  return {
+    refs,
+    floatingStyles,
+    context,
+    getFloatingProps,
+    getReferenceProps: () => {
+      return {...referenceProps, onClick, onFocus, onKeyDown};
+    },
+    isOpen,
+    setIsOpen: manageOpen,
+    arrowRef,
+  };
+};
```

Comment:
I love what this thing does, but I think the API would be simpler if we return `FloatingWidget` here.
From the both this widget's story as its use in the DatePicker, I understand things like `floatStyles`, `context` and `arrowRef` are passed back into the `FloatingWidget` verbatim.

If we define `FloatingWidget` inside the hook, those can be closures.

I'm racking my brain if there is a way to make the event handling fool[1] proof, but I don't see anything obvious.

[1] with fool I mean myself as a possible future user of this component, of course.

Reply:
I think I considered that, but couldn't figure out how to wire it all together, but I'll revisit this, it might be obvious now instead of some messed up in-between state while I was developing it. The floating stuff had some requirements about direct children, but *probably* we can manage that now.